test_integration/test_utils.go

```diff
@@ -0,0 +1,362 @@
+				for placeholder, newImage := range imageReplacements {
+					if imageNode.Value == placeholder {
+						imageNode.Value = newImage
+						break
+					}
+				}
+				break
+			}
+		}
+	}
+
+	return nil
+}
+
+func buildDockerImage(dockerfilePath, imageName, imageTag string) error {
+	cli, err := dc.NewClientWithOpts(dc.FromEnv)
+	if err != nil {
+		return fmt.Errorf("failed to build docker image %q: failed to create Docker client: %v", imageName, err)
+	}
+
+	// Get the directory containing the Dockerfile
+	contextDir := filepath.Dir(dockerfilePath)
+
+	// Create a tar archive of the build context
+	tar, err := archive.TarWithOptions(contextDir, &archive.TarOptions{})
+	if err != nil {
+		return fmt.Errorf("failed to build docker image %q: failed to create tar archive: %v", imageName, err)
+	}
+	defer tar.Close()
+
+	// Set build options
+	opts := types.ImageBuildOptions{
+		Dockerfile: filepath.Base(dockerfilePath),
+		Tags:       []string{fmt.Sprintf("%s:%s", imageName, imageTag)},
+		Remove:     true,
+	}
+
+	// Build the image
+	resp, err := cli.ImageBuild(context.Background(), tar, opts)
+	if err != nil {
+		return fmt.Errorf("failed to build docker image %q: failed to build image: %v", imageName, err)
+	}
+	defer resp.Body.Close()
+
+	// Print build output
+	_, err = io.Copy(os.Stdout, resp.Body)
+	if err != nil {
+		return fmt.Errorf("failed to build docker image %q: failed to read build output: %v", imageName, err)
+	}
+
+	fmt.Printf("Image %s:%s built successfully\n", imageName, imageTag)
+	return nil
+}
+
+func deleteDockerTestImagesAndPrune(suffix string) error {
+	// Create a new Docker client
+	ctx := context.Background()
+	cli, err := dc.NewClientWithOpts(dc.FromEnv)
+	if err != nil {
+		return fmt.Errorf("deleting docker images: failed to create Docker client: %v", err)
+	}
+	defer cli.Close()
+
+	// List all imagesList
+	imagesList, err := cli.ImageList(ctx, image.ListOptions{All: true})
+	if err != nil {
+		return fmt.Errorf("deleting docker images: failed to list imagesList: %v", err)
+	}
+
+	// Iterate through the imagesList and delete those that match the suffix
+	for _, currentImage := range imagesList {
+		for _, tag := range currentImage.RepoTags {
+			if strings.HasSuffix(tag, suffix) {
+				fmt.Printf("Deleting currentImage: %s with ID: %s\n", tag, currentImage.ID)
+				_, err := cli.ImageRemove(context.Background(), currentImage.ID, image.RemoveOptions{Force: true, PruneChildren: true})
+				if err != nil {
+					fmt.Printf("deleting docker images: Failed to delete currentImage %s: %v\n", tag, err)
+				}
+				// Break after deleting to avoid trying to delete the same currentImage multiple times
+				break
+			}
+		}
+	}
+
+	pruneReport, err := cli.ImagesPrune(ctx, filters.NewArgs())
+	if err != nil {
+		return fmt.Errorf("deleting docker images: failed to prune images: %v", err)
+	}
+	fmt.Printf(
+		"Pruned %d images, reclaimed space: %d bytes\n",
+		len(pruneReport.ImagesDeleted),
+		pruneReport.SpaceReclaimed,
+	)
```

Comment:
I think the `testing.M` has a logger, perhaps we should use it?

Reply:
Unfortunately `testing.M` does not have a logger (that I am aware of), only `testing.T` have. Thats why I ended up using `Printf `. 
I was thinking of using the `log` package in multiple places in the tests, which would include `TestMain`

Reply:
+1 to using a dedicated log package